runtime-light/coroutine/shared-task.h

```diff
@@ -0,0 +1,361 @@
+  //                      values are of type 'shared_task_impl_::shared_task_waiter_t'.
+  //                      indicates that the coroutine has been started.
+  void *m_waiters{std::addressof(m_waiters)};
+};
+
+template<typename promise_type>
+class awaiter_base_t {
+  enum class state : uint8_t { init, suspend, end };
+  state m_state{state::init};
+
+protected:
+  std::coroutine_handle<promise_type> m_coro;
+  shared_task_impl_::shared_task_waiter_t m_waiter{};
+
+public:
+  constexpr explicit awaiter_base_t(std::coroutine_handle<promise_type> coro) noexcept
+    : m_coro(coro) {}
+
+  awaiter_base_t(awaiter_base_t &&other) noexcept
+    : m_state(std::exchange(other.m_state, state::end))
+    , m_coro(std::exchange(other.m_coro, {}))
+    , m_waiter(std::exchange(other.m_waiter, {})) {}
+
+  awaiter_base_t(const awaiter_base_t &other) = delete;
+  awaiter_base_t &operator=(const awaiter_base_t &other) = delete;
+  awaiter_base_t &operator=(awaiter_base_t &&other) = delete;
+
+  ~awaiter_base_t() {
+    if (m_state == state::suspend) {
+      cancel();
+    }
+  }
+
+  constexpr bool await_ready() const noexcept {
+    php_assert(m_state == state::init && m_coro);
+    return m_coro.promise().ready();
+  }
+
+  constexpr bool await_suspend(std::coroutine_handle<> awaiter) noexcept {
+    m_state = state::suspend;
+    m_waiter.m_continuation = awaiter;
+    return m_coro.promise().suspend_awaiter(m_waiter);
+  }
+
+  constexpr void await_resume() noexcept {
+    m_state = state::end;
+  }
+
+  constexpr bool resumable() const noexcept {
+    return m_coro.promise().ready();
+  }
+
+  constexpr void cancel() noexcept {
+    m_state = state::end;
+    m_coro.promise().cancel_awaiter(m_waiter);
+  }
+};
+
+} // namespace shared_task_impl_
+
+template<typename T>
+struct shared_task_t final {
+  template<std::same_as<T> F>
+  struct promise_non_void_t;
+  struct promise_void_t;
+
+  using promise_type = std::conditional_t<!std::is_void_v<T>, promise_non_void_t<T>, promise_void_t>;
+
+  constexpr explicit shared_task_t(std::coroutine_handle<> coro) noexcept
+    : m_haddress(coro.address()) {}
+
+  constexpr shared_task_t(const shared_task_t &other) noexcept
+    : m_haddress(other.m_haddress) {
+    if (m_haddress != nullptr) [[likely]] {
+      std::coroutine_handle<promise_type>::from_address(m_haddress).promise().add_ref();
+    }
+  }
+
+  constexpr shared_task_t(shared_task_t &&other) noexcept
+    : m_haddress(std::exchange(other.m_haddress, nullptr)) {}
+
+  shared_task_t &operator=(const shared_task_t &other) noexcept {
+    if (m_haddress != other.m_haddress) [[likely]] {
+      destroy();
+      m_haddress = other.m_haddress;
+      if (m_haddress) [[likely]] {
+        std::coroutine_handle<promise_type>::from_address(m_haddress).promise().add_ref();
+      }
+    }
+    return *this;
+  }
+
+  shared_task_t &operator=(shared_task_t &&other) noexcept {
+    if (this != std::addressof(other)) [[likely]] {
```

Comment:
Why this condition isn't symmetric condition in `shared_task_t &operator=(const shared_task_t &other)`?